design/control-data-plane-separation/design.md

```diff
@@ -0,0 +1,770 @@
+Rather than sending the entire nginx configuration on every update, we could update only what has changed, such as
+endpoints, servers, or locations. We could use the [Nginx Controller API][controller], the [N+ API][nplus], and
+the [xDS API][xds] as inspiration for the design.
+
+Benefits:
+
+- We can configure nginx on a more granular level.
+- We have control over the design and implementation.
+- We can add dynamic configuration wherever possible
+
+Drawbacks:
+
+- Time. It would take a lot of time to develop, implement, and maintain this agent.
+- We would add yet another nginx agent to the ecosystem.
+
+**Approach 3: Use the xDS API and build an agent that translates xDS to nginx config**
+
+The xDS API was developed for Envoy, however there is a [CNCF xDS API Working Group][xds-wg] that is working to create a
+universal data plane API based on the xDS APIs.
+
+Benefits:
+
+- We can participate xDS Working Group and help drive the evolution of the APIs
+- We don't have to design our own API
+- We could focus on the data plane and bring in an xDS control plane
+- An agent that translates xDS to nginx config could be useful for other projects
+
+Drawbacks:
+
+- Steep learning curve. We would need to learn xDS APIs
+- Translating xDS to nginx config is not a trivial task. Some features will not map easily to nginx config.
+- We would either need to completely re-write our control plane or adopt an xDS control plane
+
+### Modify nginx-agent
+
+We could modify the nginx agent to meet our needs. This would require us to maintain a fork of the nginx agent. I see
+this option as a last resort that we can decide to adopt if we hit a roadblock with the nginx agent.
+
+[controller]: https://docs.nginx.com/nginx-controller/api/overview/
+
+[nplus]: http://nginx.org/en/docs/http/ngx_http_api_module.html
+
+[xds]: https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol
+
+[xds-wg]: https://github.com/cncf/xds
+
+## Deployment Architecture
+
+![Deployment architecture](deployment-architecture.png)
+
+* _Control Plane Deployment_: The control plane is a Kubernetes Deployment with one container running the NKG
+  controller. Initially, the control plane will be limited to a single Pod. Once we add leader election, the control
+  plane will be able to scale. The control plane will perform the same functions as it does today, but instead of
+  configuring nginx by writing files to a shared volume, it will send the configuration to the agent via gRPC.
+* _Control Plane Service_: Exposes the control plane via a Kubernetes Service of type `ClusterIP`. The data plane will
+  use the DNS name of the Service to connect to the control plane.
+* _Data Plane DaemonSet/Deployment_: A user can deploy the data plane as either a DaemonSet or Deployment. The data
+  plane contains a single container running both the agent and nginx processes. The agent will download the
```

Comment:
do we need a separate design for this?

For example:

Will the agent be PID 1 that will also start NGINX?

What will happen if NGINX master fails (terminates) ? What will happen if the agent fails?

How to gracefully shutdown NGINX?

When should k8s send traffic to NGINX (readiness probe)? When should k8s restart the pod (liveliness probe)?